### PR TITLE
BlockOperator on device

### PR DESCRIPTION
**Pull request by dylan-copeland**
This PR fixes (or avoids as a workaround) the issues of `BlockOperator` on device. For the current master version, running ex5 on device results in a failed assertion (`Cannot register an alias` in `MemoryManager::Register_`). I am not familiar with the `MemoryManager` class and how it should work on device, so I will defer to the cuda experts, but it appears to me that the usage of `BlockVector` in `BlockOperator` with reuse of `x.GetData()` and `y.GetData()` is not compatible with the current implementation of memory management on device. The best solution may be to resolve this, but instead I have implemented changes to `BlockOperator` and `BlockDiagonalPreconditioner` to use multiple `Vector` instances instead of `BlockVector`.

The obvious disadvantage of this new approach is the use of more memory (allocating two arrays of vectors, using memory `2N` where `N` is the total size of the `BlockOperator`) and copying of data between these vectors and the input and output vectors `x` and `y`. If this can be avoided by a change to `MemoryManager`, that may be better. However, the additional memory is modest, and the copying does not appear to be very slow. For example, running ex5 with `-m ../data/beam-hex.mesh -o 1 -pa` is 40x faster with `-d cuda`, so the copying of data does not prevent good performance.

Another option would be to switch between the old version and new version in `BlockOperator::Mult` depending on whether the input or output vector uses the device, so that the memory allocation and data copying is done only when using the device. This would preserve the old version for CPU runs. If the alias issue is resolved later, it would be easy to eliminate the new version.

If someone thinks there is a better way to implement this, this PR can be deleted, but for now it is the only way to get fast performance for ex5 (and other applications with `BlockOperator`) with cuda. Reviewers, if you think this approach is good and should be pulled into master, then I will continue working on the PR to make it ready for review. Several things remain to be done, e.g. `BlockOperator::MultTranspose` and ex5p. Thanks in advance for any feedback.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1468](https://github.com/mfem/mfem/pull/1468) | @dylan-copeland | @tzanio | @pazner + @camierjs + @mlstowell | 07/04/20 | 07/22/20 | 07/27/20 | |

### CODE REVIEW

**Comment by v-dobrev** [comment]
The right way to do this without copying the data is to use the methods https://github.com/mfem/mfem/blob/0edf8ffe4a2968fe8d7466b3c95e818e1b62a731/linalg/vector.hpp#L146-L151

to create/update "wrapping" `Vector`s pointing to sub-`Vector`s. These methods should work regardless of where the `base` data is, on CPU or GPU.

One issue that comes up here is the "wrapping" of the input `const Vector` -- one simple option is to explicitly use `const_cast` to remove the `const`. This should be fine as long as we make sure that the "wrapping" `Vector` is not actually modified.

I think, it is natural to create a version of `BlockVector::Update` that takes a `Vector &` as input to create a "wrapping" `BlockVector`. It can use the above `MakeRef` methods, instead of `Vector::NewDataAndSize` which needs a host pointer (e.g. obtained from `Vector::GetData`). Then we can simply update `BlockOperator` to call this new version of `BlockVector::Update` instead of the one using a row pointer.

**Comment by tzanio** [comment]
LGTM

**Comment by dylan-copeland** [comment]
Thanks @v-dobrev for the suggestions. I believe commit 039a04b3e2d8cb2294e3607819a373e5e4b481d6 implements what you described, but it still runs into the failed assertion `!alias` on mem_manager.cpp:669 (`Cannot register an alias!`). This happens for the `xblock.GetBlock(0)` (input) vector for block (0,0) of the `BlockOperator` in ex5, when `ElementRestriction::Mult` is called and the input vector is `Read()`. Please let me know if you see what is wrong (the alias issue is still cryptic to me).

**Comment by stefanhenneking** [comment]
The recent changes to this PR were made based on @v-dobrev explanation in #1578.

PR TODOs:
- [x] BlockOperator::Mult
- [ ] BlockOperator::MultTranspose
- [x] BlockDiagonalPreconditioner::Mult
- [ ] BlockDiagonalPreconditioner::MultTranspose
- [x] Enable support for device in ex5

Timings obtained on [Ray](https://hpc.llnl.gov/hardware/platforms/Ray) (ex5 MINRES solver):
`ex5 -m ../data/beam-hex.mesh -pa` : `7.26 sec`
`ex5 -m ../data/beam-hex.mesh -pa -d cuda` : `0.48 sec`

`ex5 -m ../data/star.mesh -pa` : `6.49 sec`
`ex5 -m ../data/star.mesh -pa -d cuda` : `0.30 sec`

**Comment by stefanhenneking** [comment]
@dylan-copeland If you think this looks good, we can mark it as `ready-for-review`.

**Comment by dylan-copeland** [comment]
@stefanhenneking Thank you for salvaging this PR! I added `-d cuda` support for ex5p, and both ex5 and ex5p work in serial, but ex5p in parallel is still broken.

**Comment by stefanhenneking** [comment]
Addition:
- [x] Enable support for device in ex5p

Timings obtained on [Ray](https://hpc.llnl.gov/hardware/platforms/Ray) (ex5p MINRES solver):
`mpirun -np 4 ex5p -m ../data/beam-hex.mesh -r 2 -pa` : `55.6 sec`
`mpirun -np 4 ex5p -m ../data/beam-hex.mesh -r 2 -pa -d cuda` : `5.76 sec`

`mpirun -np 4 ex5p -m ../data/star.mesh -r 2 -pa` : `2.25 sec`
`mpirun -np 4 ex5p -m ../data/star.mesh -r 2 -pa -d cuda` : `2.42 sec`

**Comment by dylan-copeland** [comment]
@stefanhenneking I was able to get better results, for the same tests on same machine (ray):

`srun -n 4 -p pdebug ./ex5p -m ../data/beam-hex.mesh -r 2 -pa`: 56.6311s
`srun -n 4 -p pdebug ./ex5p -m ../data/beam-hex.mesh -r 2 -pa -d cuda`: 1.36063s

`srun -n 4 -p pdebug ./ex5p -m ../data/star.mesh -r 2 -pa`: 2.15889s
`srun -n 4 -p pdebug ./ex5p -m ../data/star.mesh -r 2 -pa -d cuda`: 0.435838s

The difference is `mpirun` vs `srun` or `lrun`. I found in the past that `mpirun` only uses 1 of the 4 GPUs (this can be seen by using `nvidia-smi`). So this PR is working very well.

**Comment by dylan-copeland** [comment]
This PR is ready for review now, and the only remaining issue @stefanhenneking and I see is a matter of style, that `ex5(p)` requires some special code to work properly on GPUs. Ideally, it would be nice to eliminate these lines, to keep the example as simple as possible and easier to read, but for the current implementation they seem to be necessary. Suggestions for simplification are welcome.

**Comment by stefanhenneking** [comment]
@dylan-copeland: I'd like to reference #1578 for the latest issue described there; it also affects the `BlockOperator`, and thus a fix _may_ be needed in the `BlockOperator`. If `BlockOperator` does need modification to resolve it, I think this would be a rather small addition and could be done during the review or post-review in this PR, based on the discussion in #1578.

**Comment by stefanhenneking** [comment]
The latest commit is a fix to the issue with dangling aliases in the memory manager, described in #1578.

**Comment by pazner** [comment]
I recall encountering some similar issues when working on Stokes on GPU. @camierjs, do you remember if this is the same issue?

**Comment by tzanio** [comment]
LGTM

**Comment by stefanhenneking** [comment]
@dylan-copeland @mlstowell @pazner @camierjs @v-dobrev 
I have made a small change to the comments regarding the `Destroy` statements. As discussed, this PR should be ready now, and we continue discussion on the issue of dangling pointers in #1578 and a new branch (tbd).

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: CHANGELOG [hunk 1]
```diff
@@ -58,6 +58,8 @@ Improved GPU capabilities
   compute a global sparse matrix. All integrators supported by element assembly
   are also supported by full assembly. See the '-fa' option in Example 9.
 
+- Added support for BlockOperator on GPU.
```
Comment (tzanio) [inline CHANGELOG:61]:
```suggestion
- Added support for BlockOperator on GPU. See the updated Example 5.
```

**Comment by tzanio** [comment]
Thanks @dylan-copeland and @stefanhenneking !

**Pull request merged**